### PR TITLE
Use importlib.metadata to read requirements

### PR DESCRIPTION
**Pull request by DMRobertson**
Part of #11537. We will have to use `importlib.metadata` to read our dependencies at runtime once we have moved to a declarative config. I do this now to ease the transition.

Before this change, python_dependencies has four confused purposes:

1. setup.py evaluates this to define matrix-synapse's dependencies and optional
   dependencies.
2. The twisted olddeps CI job edit this file in-place so that we install the oldest
   version of all dependencies. Similarly, the twisted trunk CI job patches this file
   to use twisted trunk from github.
3. This data is made available at runtime for the application to check dependencies
   are present and correctly versioned.
4. Finally, the script can be executed directly to print out all dependencies
   (though this lacks the extras grouping information). This is mentioned in the
   upgrade notes for 0.7.0 and nowhere else.

Purposes (1) and (2) have no right belonging in application code. I am relocating and reimplementing purpose (3) here. I propose nuking purpose (4) as well while we're at it---again, this is not something the application source code should be concerned with.

I've tested this with some dirty mocked test cases, and also:

```
$ pip install 'twisted<18.9' 'twisted>16'
Collecting twisted<18.9
  Downloading Twisted-18.7.0.tar.bz2 (3.1 MB)
     |████████████████████████████████|
 ...
    Running setup.py install for twisted ... done
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
matrix-synapse 1.53.0 requires Twisted>=18.9.0, but you have twisted 18.7.0 which is incompatible.
Successfully installed PyHamcrest-2.0.3 twisted-18.7.0

$ python -m synapse.app
ERROR:root:Need Twisted>=18.9.0, but got Twisted==18.7.0
Missing Requirements: "Twisted"
To install run:
    pip install --upgrade --force "Twisted"
```

### CODE REVIEW

**Comment by DMRobertson** [comment]
Sytest is not happy. Grr.

**Comment by DMRobertson** [comment]
#12089 seems to have hit the same set of sytest failures(?)

**Comment by anoadragon453** [comment]
Sytest failures are affecting develop as well, yes: https://github.com/matrix-org/synapse/runs/5356507389?check_suite_focus=true

File: tests/util/test_check_dependencies.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        raise NotImplementedError()
+
+    def read_text(self, filename):
+        raise NotImplementedError()
+
+
+old = DummyDistribution("0.1.2")
+new = DummyDistribution("1.2.3")
+
+# could probably use stdlib TestCase --- no need for twisted here
```
Comment (squahtx) [inline tests/util/test_check_dependencies.py:32]: I've got no strong opinions on which `TestCase` class we use.

**Review by squahtx — APPROVED** [review]
LGTM!

Just to confirm, this PR only touches purpose (3)?

File: synapse/util/check_dependencies.py [hunk 2]
```diff
@@ -29,79 +27,103 @@ def dependencies(self):
...
+    elif extra in EXTRAS:
+        dependencies = _dependencies_for_extra(extra)
+    else:
+        raise ValueError(f"Synapse does not provide the feature '{extra}'")
 
-    # first check if the markers specify that this requirement needs installing
-    if req.marker is not None and not req.marker.evaluate():
-        # not required for this environment
-        return
+    deps_needed = []
```
Comment (squahtx) [inline synapse/util/check_dependencies.py:110]: nit: `needed` is a tad misleading. `deps_unfulfilled` maybe?

Comment (DMRobertson) [reply synapse/util/check_dependencies.py:110]: Can do. This came from the original implementation, but finding the right words for all of this stuff is tricky.

File: synapse/util/check_dependencies.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    def dependencies(self) -> Iterable[str]:
+        for i in self.args[0]:
+            yield '"' + i + '"'
+
+
+EXTRAS = {
+    v
+    for k, v in metadata.distribution("matrix-synapse").metadata.items()
+    if k == "Provides-Extra"
+}
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/util/check_dependencies.py:34]: That seems neater---will check this out!

[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/util/check_dependencies.py:34]: Seems kosher to me: https://importlib-metadata.readthedocs.io/en/latest/api.html#importlib_metadata.PackageMetadata.get_all

**Review by squahtx — APPROVED** [review]

**Comment by DMRobertson** [comment]
Needs a linter script run, but I'll put it to auto-merge after that.

**Comment by DMRobertson** [comment]
This change didn't mark `importlib_metadata` as a requirement! We are fortunate that the backport is pulled in by other dependencies.

**Pull request merged**